include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -260,4 +263,197 @@ def YieldOp : VerifOp<"yield", [
+    of the instanciated `hw.module` but inside of this region. Assertions/Assumptions 
+    defined within the instanciated module will also be used for verification. 
+    The region will then be converted into btor2, SystemVerilog, or used for verification 
+    directly in circt-bmc. The operations in this region are ignored during regular SystemVerilog 
+    emission.   
+      
+    Additional information about how the test should be conducted can be passed 
+    using the `k` attribute, which defines the bounds of the test.
+  }];
+
+  let arguments = (ins SymbolNameAttr:$sym_name, APIntAttr:$k);
+
+  let regions = (region SizedRegion<1>:$test_region);
+
+  let assemblyFormat = [{
+    $sym_name `(``k` `=` $k `)` attr-dict-with-keyword
+    $test_region
+  }];
+
+  let extraClassDeclaration = [{
+    /// Implement RegionKindInterface.
+    static RegionKind getRegionKind(unsigned index) { 
+      return RegionKind::Graph;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Verification Ops
+//===----------------------------------------------------------------------===//
+
+def SymbolicInputOp : VerifOp<"symbolic_input", [HasParent<"verif::FormalOp">]>{
+  let summary = "declare a symbolic input for formal verification";
+  let description = [{
+    This operation declares a symbolic input that can then be used to reason 
+    about `hw.instance` inputs in a symbolic manner in future assertions and 
+    assumptions. The result type must be explicitly marked. The resulting value 
+    is a new non-deterministic value for every implicit clock cycle. 
+  }];
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+def ConcreteInputOp : VerifOp<"concrete_input", [
+  AllTypesMatch<["init", "update"]>, HasParent<"verif::FormalOp">
+]> {
+  let summary = "declare a concrete input for formal verification";
+  let description = [{
+    This operation declares a concrete input that can then be used in the 
+    reasoning surrounding symbolic inputs, allowing for a form of concolic 
+    reasoning to take place in a `verif.formal` block. Concrete inputs are 
+    defined by an initial value and an update SSA value. This is equivalent 
+    to a register with an implicit clock, which is initialized with the init 
+    value and obtains a new value every implicit clock tick from the update 
+    value.
+  }];
+
+  let arguments = (ins AnyType:$init, AnyType:$update);
+  let results = (outs AnyType:$result);
+
+  let assemblyFormat = [{
+    $init `,` $update attr-dict `:` 
+    functional-type(operands, results)
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Contract Ops
+//===----------------------------------------------------------------------===//
+
+def ContractOp : VerifOp<"contract", [Pure, HasParent<"hw::HWModuleOp">]>{
```

Comment:
In the ops current form, the user can use arbitrary SSA values from the hw module body in this op which means the contract and design are not entirely separated. Did you consider isolating this op from above and providing the module ports as block arguments within the region? Additionally, the contract could live next to a module instead of inside of it which could be useful for external modules, but not sure if that's needed. Just wondering if you thought about this?